src/platform/tls_openssl.c

```diff
@@ -0,0 +1,3437 @@
+    _In_ CXPLAT_SEC_CONFIG* SecurityConfig,
+    _In_reads_(KeyCount) QUIC_TICKET_KEY_CONFIG* KeyConfig,
+    _In_ uint8_t KeyCount
+    )
+{
+    CXPLAT_DBG_ASSERT(KeyCount >= 1); // Only support 1, ignore the rest for now
+    UNREFERENCED_PARAMETER(KeyCount);
+
+    if (SecurityConfig->Flags & QUIC_CREDENTIAL_FLAG_CLIENT) {
+        return QUIC_STATUS_NOT_SUPPORTED;
+    }
+
+    if (SecurityConfig->TicketKey == NULL) {
+        SecurityConfig->TicketKey =
+            CXPLAT_ALLOC_NONPAGED(sizeof(QUIC_TICKET_KEY_CONFIG), QUIC_POOL_TLS_TICKET_KEY);
+        if (SecurityConfig->TicketKey == NULL) {
+            QuicTraceEvent(
+                AllocFailure,
+                "Allocation of '%s' failed. (%llu bytes)",
+                "QUIC_TICKET_KEY_CONFIG",
+                sizeof(QUIC_TICKET_KEY_CONFIG));
+            return QUIC_STATUS_OUT_OF_MEMORY;
+        }
+    }
+
+    CxPlatCopyMemory(
+        SecurityConfig->TicketKey,
+        KeyConfig,
+        sizeof(QUIC_TICKET_KEY_CONFIG));
+
+    SSL_CTX_set_tlsext_ticket_key_evp_cb(
+        SecurityConfig->SSLCtx,
+        CxPlatTlsOnSessionTicketKeyNeeded);
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+//
+// @brief OpenSSL BIO callback to free application-specific auxiliary data.
+//
+// This function is registered as a callback with a BIO and is called when
+// the BIO is being freed. If the operation is @c BIO_CB_FREE, it retrieves
+// the custom @c AUX_DATA structure previously stored using
+// @c BIO_set_app_data, releases its internal allocations, and clears the
+// pointer from the BIO.
+//
+// This is used in the QUIC TLS layer to clean up transport parameters,
+// peer transport parameters, and the auxiliary data struct itself.
+//
+// @param[in] b
+//     Pointer to the BIO being freed.
+// @param[in] oper
+//     Operation code. Only @c BIO_CB_FREE triggers cleanup.
+// @param[in] argp
+//     Unused parameter.
+// @param[in] len
+//     Unused parameter.
+// @param[in] argi
+//     Unused parameter.
+// @param[in] arg1
+//     Unused parameter.
+// @param[in] ret
+//     Return value passed through by OpenSSL.
+// @param[in] Processed
+//     Unused parameter.
+//
+// @return The value of @p ret, unmodified.
+//
+static long FreeBioAuxData(BIO *B, int Oper,
+                           const char *ArgP,
+                           size_t Len,
+                           int ArgInt,
+                           long ArgLong,
+                           int Ret,
+                           size_t *Processed)
+{
+    struct AUX_DATA *AData;
+
+    UNREFERENCED_PARAMETER(*ArgP);
+    UNREFERENCED_PARAMETER(Len);
+    UNREFERENCED_PARAMETER(ArgInt);
+    UNREFERENCED_PARAMETER(ArgLong);
+    UNREFERENCED_PARAMETER(Processed);
+
+    if (Oper == BIO_CB_FREE) {
+        AData = BIO_get_app_data(B);
+        CXPLAT_FREE(AData->Tp, QUIC_POOL_TLS_TRANSPARAMS);
+        CXPLAT_FREE(AData->PeerTp, QUIC_POOL_TLS_TRANSPARAMS);
+        CXPLAT_FREE(AData, QUIC_POOL_GENERIC);
```

Comment:
Let's create a new tag for the OpenSSL AuxData allocation so allocations can be properly attributed during memory leak investigations. See quic_platform.h to define a new tag there.